### PR TITLE
[SPARK-31502][SQL][DOCS] Document identifier in SQL Reference

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
Document identifier in SQL Reference

### Why are the changes needed?
make SQL Reference complete

### Does this PR introduce any user-facing change?
Yes
Screen Shot 2020-04-23 at 11 14 10 PM

Screen Shot 2020-04-23 at 11 32 32 PM

### How was this patch tested?
Manually build and check

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @cloud-fan

**Comment by huaxingao** [comment]
also cc @maropu

**Comment by huaxingao** [comment]
@cloud-fan @maropu 
One more to review. Thank you very much!

**Comment by maropu** [comment]
Ur, I forgot to submit my reviews last night... some comments might be stale...

File: docs/sql-ref-identifier.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+---
+
+### Description
+
+An identifier is a string used to identify a database object such as a table, view, schema, column, etc. Spark SQL has regular identifiers and delimited identifiers, which are enclosed within backticks. Both regular identifiers and delimited identifiers are case insensitive.
```
[the remark this answers is not in the document]
Comment (huaxingao) [reply docs/sql-ref-identifier.md:24]: Regular identifiers and delimited identifiers are common words. By standard, delimited identifiers are case sensitive. 
https://en.wikibooks.org/wiki/SQL_Dialects_Reference/Data_structure_definition/Delimited_identifiers

**Comment by maropu** [comment]
Looks fine now cc: @srowen @cloud-fan

**Review by maropu — APPROVED** [review]

File: docs/sql-ref-identifier.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+---
+
+### Description
+
+An identifier is a string used to identify a database object such as a table, view, schema, column, etc. Spark SQL has regular identifiers and delimited identifiers, which are enclosed within backticks. When `spark.sql.caseSensitive` is set to false (default behavior since Spark 2.4), both regular identifiers and delimited identifiers are case-insensitive.
```
Comment (cloud-fan) [inline docs/sql-ref-identifier.md:24]: `spark.sql.caseSensitive` is an internal config. Shall we ignore it and just say Spark is case insensitive in the doc?

File: docs/sql-ref-identifier.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+### Syntax
+
+#### Regular Identifier
+
+{% highlight sql %}
+{ letter | digit | '_' } [ , ... ]
```
Comment (cloud-fan) [inline docs/sql-ref-identifier.md:31]: `[ , ... ]` -> `[ c ... ]`? IIRC only the first letter needs to be `letter | digit | '_'`

File: docs/sql-ref-identifier.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+</dl>
+
+### Examples
+
+{% highlight sql %}
+-- This CREATE TABLE fails because of the illegal identifier name a.b
+CREATE TABLE test (a.b int);
+
+-- This CREATE TABLE works
+CREATE TABLE test (`a.b` int);
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply docs/sql-ref-identifier.md:69]: It's still beta, we don't mention anything about it in the sql reference for now.

File: docs/sql-ref-identifier.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+-- This CREATE TABLE fails
+CREATE TABLE test1 (`a`b` int);
```
Comment (cloud-fan) [inline docs/sql-ref-identifier.md:72]: this should fail the parser?

File: docs/sql-ref-identifier.md (hunk 2 above)
Comment (huaxingao) [reply docs/sql-ref-identifier.md:24]: Sure. Will update.

File: docs/sql-ref-identifier.md (hunk 5 above)
Comment (huaxingao) [reply docs/sql-ref-identifier.md:72]: Yes. This throws ParseException. Do you want me to change L71 to 
```-- This CREATE TABLE fails with ParseException```?

Comment (cloud-fan) [reply docs/sql-ref-identifier.md:72]: Can we include the error message in the example?

Comment (huaxingao) [reply docs/sql-ref-identifier.md:74]: Updated. Thanks!

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by huaxingao** [comment]
Thank you all!

**Pull request closed without merging**